File: poker_ai/equity.py
"""Équité (probabilité de gain) par simulation Monte-Carlo vectorisée."""
import numpy as np

from .evaluator import evaluate_hands


def equity_vs_random(hole, board, n_sims, rng):
    """Équité de `hole` (2 cartes) + `board` (0/3/4/5 cartes) contre une main
    adverse aléatoire, board complété aléatoirement. Renvoie un float 0..1."""
    known = list(hole) + list(board)
    known_set = set(known)
    deck = np.array([c for c in range(52) if c not in known_set], dtype=np.int64)

    need = 5 - len(board)
    k = 2 + need  # 2 cartes adverses + fin du board

    # Un tirage sans remise par simulation : k premiers indices d'une permutation
    perm = np.argsort(rng.random((n_sims, deck.size)), axis=1)[:, :k]
    draw = deck[perm]
    opp = draw[:, :2]
    runout = draw[:, 2:]

    mine = np.broadcast_to(np.array(known, dtype=np.int64), (n_sims, len(known)))
    my7 = np.column_stack([mine, runout])
    if board:
        board_arr = np.broadcast_to(np.array(board, dtype=np.int64), (n_sims, len(board)))
        opp7 = np.column_stack([opp, board_arr, runout])
    else:
        opp7 = np.column_stack([opp, runout])

    my_scores = evaluate_hands(my7)
    opp_scores = evaluate_hands(opp7)
    wins = (my_scores > opp_scores).sum()
    ties = (my_scores == opp_scores).sum()
    return float((wins + 0.5 * ties) / n_sims)

File: native/src/equity.rs
//! Équité Monte-Carlo contre main aléatoire — parité statistique avec
//! poker_ai/equity.py (même espérance, générateur différent).

use crate::eval::evaluate7;
use crate::rng::Rng;

pub fn equity_vs_random(hole: &[u8; 2], board: &[u8], n_sims: usize, rng: &mut Rng) -> f64 {
    let mut known = [false; 52];
    known[hole[0] as usize] = true;
    known[hole[1] as usize] = true;
    for &c in board {
        known[c as usize] = true;
    }
    let mut deck: Vec<u8> = (0..52u8).filter(|&c| !known[c as usize]).collect();

    let need = 5 - board.len();
    let k = 2 + need;

    let mut my7 = [0u8; 7];
    my7[0] = hole[0];
    my7[1] = hole[1];
    my7[2..2 + board.len()].copy_from_slice(board);

    let mut opp7 = [0u8; 7];
    opp7[2..2 + board.len()].copy_from_slice(board);

    let mut wins = 0u32;
    let mut ties = 0u32;
    for _ in 0..n_sims {
        rng.partial_shuffle(&mut deck, k);
        opp7[0] = deck[0];
        opp7[1] = deck[1];
        for i in 0..need {
            my7[2 + board.len() + i] = deck[2 + i];
            opp7[2 + board.len() + i] = deck[2 + i];
        }
        let ms = evaluate7(&my7);
        let os = evaluate7(&opp7);
        if ms > os {
            wins += 1;
        } else if ms == os {
            ties += 1;
        }
    }
    (wins as f64 + 0.5 * ties as f64) / n_sims as f64
}
